zerver/lib/api.py

```diff
@@ -0,0 +1,334 @@
+            except UnicodeDecodeError:
+                raise JsonableError(_("Malformed payload"))
+        elif request_var_name in request.POST:
+            value_to_parse = request.POST[request_var_name]
+            processed_request_var_names.add(request_var_name)
+        elif request_var_name in request.GET:
+            value_to_parse = request.GET[request_var_name]
+            processed_request_var_names.add(request_var_name)
+        else:
+            if parameter.default is NotSpecified:
+                raise RequestVariableMissingError(request_var_name)
+            # By skipping here, we leave it to Python to use the default value
+            # of this parameter, because we cannot find the request variable in
+            # the request.
+            # This is covered by test_json, but coverage.py
+            # fails to recognize this for some reason.
+            continue  # nocoverage
+
+        try:
+            if parameter.metadata.convert_from_json:
+                param_value = json_converter.loads(value_to_parse, parameter.param_type)
+            else:
+                param_value = zulip_converter.structure(value_to_parse, parameter.param_type)
+        except orjson.JSONDecodeError:
+            # The JSON decoder should only be called when convert_from_json is True
+            assert parameter.metadata.convert_from_json
+            if parameter.metadata.argument_type_is_body:
+                raise InvalidJSONError(_("Malformed JSON"))
+            raise JsonableError(
+                _('Argument "{name}" is not valid JSON.').format(name=request_var_name)
+            )
+        except BaseException as exc:
+            errors.extend(
+                transform_error(exc, path=request_var_name, format_exception=format_exception)
+            )
+        else:
+            kwargs_to_update[parameter.param_name] = param_value
+
+    return CallInfo(
+        kwargs_to_update=kwargs_to_update,
+        processed_request_var_names=processed_request_var_names,
+        errors=errors,
+    )
+
+
+def endpoint(
+    view_func: Callable[Concatenate[HttpRequest, ParamT], ReturnT]
+) -> Callable[Concatenate[HttpRequest, ParamT], ReturnT]:
+    # Extract all the type information from the view function.
+    endpoint_info = parse_view_func_signature(view_func)
+    for func_param in endpoint_info.parameters:
+        if func_param.metadata.path_only:
+            assert (
+                func_param.default is NotSpecified
+            ), f"Path-only parameter {func_param.param_name} should not have a default value"
+        # Record arguments that should be documented so that our
+        # automated OpenAPI docs tests can compare these against the code.
+        if (
+            func_param.metadata.documentation_status is DocumentationStatus.DOCUMENTED
+            and not func_param.metadata.path_only
+        ):
+            # TODO: Move arguments_map to here once zerver.lib.request does not
```

Comment:
Same question about "TODO" as before.

Reply:
This is also the same TODO item from `has_request_variables`. We plan to work on that separately.